### PR TITLE
Non essential container state change bug fix

### PR DESCRIPTION
**Pull request by aaithal**
### Summary

Fixes the bug wherein non-essential container state changes were not being submitted to ECS

### Implementation details

Added a ticker to drain the event queue every minute. If
there are any container state changes not submitted,
the ticker ensures that they do get submitted.

### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: Yes
### Description for the changelog

Will add an entry in CHANGELOG.md

### Licensing

This contribution is under the terms of the Apache 2.0 License: Yes

### CODE REVIEW

File: agent/eventhandler/task_handler.go [hunk 1]
```diff
@@ -256,3 +279,77 @@ func (handler *TaskHandler) SubmitTaskEvents(taskEvents *eventList, client api.E
...
+	}
+}
+
+// setTaskChangeSent sets the event's task change object as sent
+func setTaskChangeSent(event *sendableEvent) {
+	if event.taskChange.Task != nil {
+		event.taskChange.Task.SetSentStatus(event.taskChange.Status)
+	}
+	for _, container := range event.taskChange.Containers {
+		container.Container.SetSentStatus(event.taskChange.Status.ContainerStatus(container.Container.GetSteadyStateStatus()))
```
Comment (richardpen) [inline agent/eventhandler/task_handler.go]: Why not just `container.Status` instead of `event.taskChange.Status.ContainerStatus(container.Container.GetSteadyStateStatus())`?

File: agent/eventhandler/task_handler_types.go [hunk 2]
```diff
@@ -71,16 +72,27 @@ func (event *sendableEvent) taskArn() string {
...
 	if tevent.Status == api.TaskStatusNone {
 		return false // defensive programming :)
 	}
-	if event.taskSent || (tevent.Task != nil && tevent.Task.GetSentStatus() >= tevent.Status) {
+	if event.taskSent {
 		return false // redundant event
 	}
+	if tevent.Task != nil && tevent.Task.GetSentStatus() >= tevent.Status {
+		for _, containerStateChange := range tevent.Containers {
+			container := containerStateChange.Container
```
Comment (richardpen) [inline agent/eventhandler/task_handler_types.go]: Please add some comments here to makes it easier to read. And this part should have a unit test to cover.

File: agent/eventhandler/task_handler.go [hunk 3]
```diff
@@ -188,57 +254,14 @@ func (handler *TaskHandler) SubmitTaskEvents(taskEvents *eventList, client api.E
 			event := eventToSubmit.Value.(*sendableEvent)
 
 			if event.containerShouldBeSent() {
```
Comment (richardpen) [inline agent/eventhandler/task_handler.go]: We probably don't need this part of code now, which can be in another pr.

File: agent/eventhandler/task_handler.go [hunk 4]
```diff
@@ -58,16 +67,73 @@ type TaskHandler struct {
...
+
+// getBatchedContainerEvents gets a list task state changes for container events that
+// have been batched and not sent beyond the drainEventsFrequency threshold
+func (handler *TaskHandler) getBatchedContainerEvents() []api.TaskStateChange {
+	handler.taskHandlerLock.RLock()
+	defer handler.taskHandlerLock.RUnlock()
+
+	var events []api.TaskStateChange
+	for taskARN, _ := range handler.tasksToContainerStates {
+		if task, ok := handler.state.TaskByArn(taskARN); ok {
```
Comment (richardpen) [inline agent/eventhandler/task_handler.go]: I'm not sure if the `else` situation will happen, but since you are checking here, we may want to remove it from the `handler.tasksToContainerStates` map if it happened. And probably a log will help.

File: agent/eventhandler/task_handler.go [hunk 5]
```diff
@@ -58,16 +67,73 @@ type TaskHandler struct {
...
+				handler.AddStateChangeEvent(event, handler.client)
+			}
+		}
+	}
+}
+
+// getBatchedContainerEvents gets a list task state changes for container events that
+// have been batched and not sent beyond the drainEventsFrequency threshold
+func (handler *TaskHandler) getBatchedContainerEvents() []api.TaskStateChange {
+	handler.taskHandlerLock.RLock()
```
Comment (richardpen) [inline agent/eventhandler/task_handler.go]: There may be a race condition where we are sending in the ticker and received the task state change also, can you move this lock inside the ticker line 110?

File: agent/eventhandler/task_handler.go [hunk 6]
```diff
@@ -58,16 +67,73 @@ type TaskHandler struct {
 	// stateSaver is a statemanager which may be used to save any
 	// changes to a task or container's SentStatus
 	stateSaver statemanager.Saver
+
+	drainEventsFrequency time.Duration
```
Comment (samuelkarp) [inline agent/eventhandler/task_handler.go]: Looks like this needs gofmt.

Comment (aaithal) [reply agent/eventhandler/task_handler.go]: Why do you think so? It is formatted code.

File: agent/eventhandler/task_handler.go (hunk 1 above)
Comment (aaithal) [reply agent/eventhandler/task_handler.go]: I missed that we have this in `ContainerStateChange`. I've modified this.

File: agent/eventhandler/task_handler.go [hunk 7]
```diff
@@ -58,16 +67,73 @@ type TaskHandler struct {
...
+func (handler *TaskHandler) startDrainEventsTicker() {
+	ticker := time.NewTicker(handler.drainEventsFrequency)
+	for {
+		select {
+		case <-handler.ctx.Done():
+			seelog.Infof("TaskHandler: Stopping periodic container state change submission ticker")
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			seelog.Info("TaskHandler tick")
```
Comment (jhaynes) [inline agent/eventhandler/task_handler.go]: Should this be `Debug` level? (and potentially line 105 too?) If we keep it at `Info`, I think we should make it more descriptive.

File: agent/eventhandler/task_handler.go [hunk 8]
```diff
@@ -58,16 +67,73 @@ type TaskHandler struct {
...
+			for _, event := range handler.getBatchedContainerEvents() {
+				seelog.Infof(
```
Comment (jhaynes) [inline agent/eventhandler/task_handler.go]: Same question as above. Should this be `Debug`?

File: agent/eventhandler/task_handler.go [hunk 9]
```diff
@@ -256,3 +279,77 @@ func (handler *TaskHandler) SubmitTaskEvents(taskEvents *eventList, client api.E
...
+	eventType string,
+	client api.ECSClient,
+	eventToSubmit *list.Element,
+	stateSaver statemanager.Saver,
+	backoff *utils.SimpleBackoff,
+	taskEvents *eventList) {
+
+	// Extract the wrapped event from the list element
+	event := eventToSubmit.Value.(*sendableEvent)
+	seelog.Infof("TaskHandler, Sending %s change: %s", eventType, event.String())
```
Comment (jhaynes) [inline agent/eventhandler/task_handler.go]: This changed from the prior if blocks from `Debug` to `Info`. I like this increase in verbosity better than some of the other `Info` level additions. However, I might flip the level on to `Debug` and log only the success message (line 309) as `Info`.

File: agent/eventhandler/task_handler.go (hunk 5 above)
Comment (aaithal) [reply agent/eventhandler/task_handler.go]: Good catch. As per offline discussion, I don't think moving the lock itself resolves the race condition. However, moving https://github.com/aws/amazon-ecs-agent/blob/dev/agent/eventhandler/task_handler.go#L114 to https://github.com/aws/amazon-ecs-agent/blob/dev/agent/eventhandler/task_handler.go#L154 does resolve it I think. I'll modify that.

File: agent/eventhandler/task_handler.go (hunk 6 above)
Comment (samuelkarp) [reply agent/eventhandler/task_handler.go]: I...have no idea what I was looking at.  Please ignore.

File: agent/eventhandler/task_handler.go (hunk 7 above)
Comment (aaithal) [reply agent/eventhandler/task_handler.go]: That's a miss on my part. It should be removed.

File: agent/eventhandler/task_handler.go (hunk 8 above)
Comment (aaithal) [reply agent/eventhandler/task_handler.go]: it makes sense for this to be `Info` as it's a trigger for submitting state changes to the backend and we'd want to know about it without having to elevate log level to debug

File: agent/eventhandler/task_handler.go (hunk 1 above)
Comment (richardpen) [reply agent/eventhandler/task_handler.go]: Looks like this is still not fixed, did you miss this in the commit?

Comment (aaithal) [reply agent/eventhandler/task_handler.go]: Seems so. Working on it.

File: agent/eventhandler/task_handler.go [hunk 10]
```diff
@@ -53,21 +59,96 @@ type TaskHandler struct {
...
+		ctx:                    ctx,
 		tasksToEvents:          make(map[string]*eventList),
 		submitSemaphore:        utils.NewSemaphore(concurrentEventCalls),
 		tasksToContainerStates: make(map[string][]api.ContainerStateChange),
 		stateSaver:             stateManager,
+		state:                  state,
+		client:                 client,
+		drainEventsFrequency:   drainEventsFrequency,
 	}
+	go taskHandler.startDrainEventsTicker()
```
[the remark this answers is not in the document]
Comment (aaithal) [reply agent/eventhandler/task_handler.go]: I think we have this pattern elsewhere as well. An object creation method doing all the things for completing the object creation doesn't seem like an unwanted thing to me. I don't have strong opinions here. I can change it if you really want me to.

File: agent/eventhandler/task_handler.go [hunk 11]
```diff
@@ -117,142 +198,218 @@ func (handler *TaskHandler) flushBatch(event *api.TaskStateChange) {
 // addEvent prepares a given event to be sent by adding it to the handler's appropriate
 // eventList and remove the entry in tasksToEvents map
 func (handler *TaskHandler) addEvent(change *sendableEvent, client api.ECSClient) {
-	handler.taskHandlerLock.Lock()
```
[the remark this answers is not in the document]
Comment (aaithal) [reply agent/eventhandler/task_handler.go]: Yeah, that's a valid race. Please look at the latest revision where this has been fixed as well.

File: agent/eventhandler/task_handler.go [hunk 12]
```diff
@@ -15,244 +15,323 @@ package eventhandler
...
 	}
 }
 
-// batchContainerEvent collects container state change events for a given task arn
-func (handler *TaskHandler) batchContainerEvent(event api.ContainerStateChange) {
-	handler.taskHandlerLock.Lock()
-	defer handler.taskHandlerLock.Unlock()
+// startDrainEventsTicker starts a ticker that periodically drains the events queue
+// by submitting state change events to the ECS backend
+func (handler *TaskHandler) startDrainEventsTicker() {
```
Comment (richardpen) [inline agent/eventhandler/task_handler.go]: nit: I'd love to move this function before `AddStateChangeEvent`.

File: agent/eventhandler/task_handler.go [hunk 13]
```diff
@@ -53,21 +59,96 @@ type TaskHandler struct {
...
+}
+
+// getBatchedContainerEvents gets a list task state changes for container events that
+// have been batched and not sent beyond the drainEventsFrequency threshold
+func (handler *TaskHandler) getBatchedContainerEvents() []api.TaskStateChange {
+	handler.lock.RLock()
+	defer handler.lock.RUnlock()
+
+	var events []api.TaskStateChange
+	for taskARN, _ := range handler.tasksToContainerStates {
```
[the remark this answers is not in the document]
Comment (richardpen) [reply agent/eventhandler/task_handler.go]: Do you want to modify this?

File: agent/eventhandler/task_handler.go [hunk 14]
```diff
@@ -15,244 +15,323 @@ package eventhandler
...
-				}
-			} else {
-				// Shouldn't be sent as either a task or container change event; must have been already sent
-				seelog.Infof("TaskHandler, Not submitting redundant event; just removing: %s", event.String())
-				taskEvents.events.Remove(eventToSubmit)
-			}
+// sendChange adds the change to the sendable events list. If the
+// event for the task is in the process of being sent, it doesn't trigger
+// the async submit state api method
+func (taskEvents *taskSendableEvents) sendChange(change *sendableEvent,
```
Comment (richardpen) [inline agent/eventhandler/task_handler.go]: Same as above, can you move this above `submitTaskEvents`.

File: agent/eventhandler/task_handler.go (hunk 12 above)
Comment (aaithal) [reply agent/eventhandler/task_handler.go]: I wanted the exported functions to be at the top of the file as that's a reasonable way to organize the code.

File: agent/eventhandler/task_handler.go (hunk 14 above)
Comment (aaithal) [reply agent/eventhandler/task_handler.go]: The order is `TaskHandler` methods and then `taskEvents` methods (exported first, non exported afterwards), which is again an organization structure that we've followed everywhere in the code base

File: agent/eventhandler/task_handler.go [hunk 15]
```diff
@@ -15,244 +15,323 @@ package eventhandler
...
+// taskSendableEvents is used to group all events for a task
+type taskSendableEvents struct {
+	// events is a list of *sendableEvents
+	events *list.List
+	// sending will check whether the list is already being handlerd
+	sending bool
+	//eventsListLock locks both the list and sending bool
+	lock sync.Mutex
+	// createdAt is a timestamp for when the event list was created
+	createdAt time.Time
```
[the remark this answers is not in the document]
Comment (aaithal) [reply agent/eventhandler/task_handler.go]: both `createdAt` and `taskARN` fields are to help with debugging/logging.

File: agent/eventhandler/task_handler.go [hunk 16]
```diff
@@ -15,244 +15,323 @@ package eventhandler
...
+	}
+	return events
 }
 
-// flushBatch attaches the task arn's container events to TaskStateChange event that
-// is being submittied to the backend
-func (handler *TaskHandler) flushBatch(event *api.TaskStateChange) {
-	handler.taskHandlerLock.Lock()
-	defer handler.taskHandlerLock.Unlock()
+// batchContainerEventUnsafe collects container state change events for a given task arn
```
[the remark this answers is not in the document]
Comment (aaithal) [reply agent/eventhandler/task_handler.go]: we're moving it up to `AddStateChangeEvent`. More details in the commit message: https://github.com/aws/amazon-ecs-agent/pull/1026/commits/d959f31b6f1a0870a4ce1f80a755a13b0d40e72e

File: agent/eventhandler/task_handler.go [hunk 17]
```diff
@@ -15,244 +15,323 @@ package eventhandler
...
-	// events is a list of *sendableEvents
-	events *list.List
-	// sending will check whether the list is already being handlerd
-	sending bool
-	//eventsListLock locks both the list and sending bool
-	eventListLock sync.Mutex
-}
+	// drainEventsFrequency is the frequency at the which unsent events batched
+	// by the task handler are sent to the backend
+	drainEventsFrequency = time.Minute
```
Comment (samuelkarp) [inline agent/eventhandler/task_handler.go]: Should we go faster than a minute?  A minute seems pretty slow to me.

File: agent/eventhandler/task_handler.go [hunk 18]
```diff
@@ -15,244 +15,323 @@ package eventhandler
...
-				}
-			} else {
-				// Shouldn't be sent as either a task or container change event; must have been already sent
-				seelog.Infof("TaskHandler, Not submitting redundant event; just removing: %s", event.String())
-				taskEvents.events.Remove(eventToSubmit)
-			}
+// sendChange adds the change to the sendable events list. If the
+// event for the task is in the process of being sent, it doesn't trigger
+// the async submit state api method
+func (taskEvents *taskSendableEvents) sendChange(change *sendableEvent,
```
Comment (samuelkarp) [inline agent/eventhandler/task_handler.go]: Can you add a bit to this comment here explaining that the actual task submission (API request) occurs asynchronously, so that this function doesn't end up causing the lock to be held for a long time?

File: agent/eventhandler/task_handler.go [hunk 19]
```diff
@@ -15,244 +15,323 @@ package eventhandler
...
+		ctx:                    ctx,
+		tasksToEvents:          make(map[string]*taskSendableEvents),
 		submitSemaphore:        utils.NewSemaphore(concurrentEventCalls),
 		tasksToContainerStates: make(map[string][]api.ContainerStateChange),
 		stateSaver:             stateManager,
+		state:                  state,
+		client:                 client,
+		drainEventsFrequency:   drainEventsFrequency,
 	}
+	go taskHandler.startDrainEventsTicker()
```
Comment (samuelkarp) [inline agent/eventhandler/task_handler.go]: I'd like a comment somewhere explaining the flow for batched events.  This is roughly what I got through reading this:

```
startDrainEventsTicker->
	getBatchedContainerEvents->
		read from tasksToContainerStates and returns TaskEvent
	AddStateChangeEvent->
		flushBatchUnsafe
```

The part that took me a while is that `AddStateChangeEvent` also calls `batchContainerUnsafe` for `ContainerEvent`s and the key difference is that `getBatchedContainerEvents` returns a `TaskEvent` even though it has "ContainerEvents" in the name.

File: agent/eventhandler/task_handler.go [hunk 20]
```diff
@@ -15,244 +15,323 @@ package eventhandler
...
+	defer handler.removeTaskEvents(taskARN)
+
+	backoff := utils.NewSimpleBackoff(submitStateBackoffMin, submitStateBackoffMax,
+		submitStateBackoffJitterMultiple, submitStateBackoffMultiple)
 
 	// Mirror events.sending, but without the need to lock since this is local
 	// to our goroutine
 	done := false
-
+	// TODO: wire in the context here. Else, we have go routine leaks in tests
```
Comment (samuelkarp) [inline agent/eventhandler/task_handler.go]: Is this TODO expected to be fixed before merging?  goroutine leaks in tests can cause flaky tests.

File: agent/eventhandler/task_handler.go [hunk 21]
```diff
@@ -15,244 +15,323 @@ package eventhandler
...
-		})
+	// Update taskEvent
+	seelog.Infof("TaskHandler: Adding event: %s", change.toString())
+	taskEvents.events.PushBack(change)
+
+	if !taskEvents.sending {
+		taskEvents.sending = true
+		go handler.submitTaskEvents(taskEvents, client, change.taskArn())
+	} else {
+		seelog.Debugf("TaskHandler: Not submitting change as the task is already being sent: %s", change.toString())
```
Comment (samuelkarp) [inline agent/eventhandler/task_handler.go]: So I think there's a race here, but the impact of the race is probably only a delay of `drainEventsFrequency`.  If we evaluate `taskEvents.sending` as true we won't cause a new goroutine to be created for `handler.submitTaskEvents`.  However, if the existing goroutine is just about done sending, it's possible it won't pick up the new event that we just added on line 263.  I think this is a reasonable tradeoff here, but it would be worth recording it somewhere that the maximum length of time we can have an unsubmitted change is really effectively `2*drainEventsFrequency`.

File: agent/eventhandler/task_handler.go (hunk 17 above)
Comment (aaithal) [reply agent/eventhandler/task_handler.go]: As per offline conversation, i'll change this to `20s`

File: agent/eventhandler/task_handler.go (hunk 20 above)
Comment (aaithal) [reply agent/eventhandler/task_handler.go]: Not with this PR :(

File: agent/eventhandler/task_handler.go (hunk 19 above)
Comment (aaithal) [reply agent/eventhandler/task_handler.go]: I'll rename `getBatchedContainerEvents` -> `taskStateChangesToSend`. Will add some comments in `startDrainEventsTicker` as well.

**Comment by aaithal** [comment]
@samuelkarp  can you take a look at the latest commit? I think I've addressed your comments. Thanks!
